packages/aws-cdk-lib/aws-s3-deployment/lib/source.ts

```diff
@@ -262,5 +258,33 @@ export class Source {
     };
   }
 
+  /**
+   * Process objects such that it escapes token output suitable for JSON output.
+   *
+   * @param scope Parent construct scope
+   * @returns Object with with tokens escaped for JSON output.
+   */
+  private static escapeTokens(scope: Construct, obj: any): any {
+    if (Token.isUnresolved(obj)) {
+      // Return tokens as numbers. This is a hack to prevent the JSON serializer to wrap this token as a string.
+      // Stack.toJsonString should take care of escaping the object output for JSON.
+      return Token.asNumber(Stack.of(scope).toJsonString(obj));
```

Comment:
I think this and than it's fine.
```suggestion
      return Token.asString(Stack.of(scope).toJsonString(obj));
```

Reply:
Any chance of conflicts with real numbers?

Reply:
To quote Amazon Q:

---

## How Number Tokens are Generated by `Token.asNumber()`

When you call `Token.asNumber(value)`, here's what happens:

### 1. **Initial Check**
If the value is already a number, it's returned as-is:
```typescript
if (typeof value === 'number') { return value; }
```

### 2. **Token Registration**
For non-number values, the token is registered in the `TokenMap` singleton:
```typescript
return TokenMap.instance().registerNumber(Token.asAny(value));
```

### 3. **Number Generation Process**
The `registerNumber` method generates a unique number through these steps:

**a) Counter Increment:**
- Uses an internal counter (`tokenCounter`) that starts at a random number (0-9)
- This randomization prevents accidental dependencies on token values between runs

**b) Special Double Encoding:**
- Calls `createTokenDouble(counter)` to create a special IEEE 754 double
- Uses a specific bit pattern in the top 16 bits to mark it as a token
- The marker pattern is `0xFBFF << 16` (can't be all 1s as that would create NaN)

**c) Bit Manipulation:**
```typescript
const buf = new ArrayBuffer(8);
const ints = new Uint32Array(buf);
ints[0] = x & 0x0000FFFFFFFF; // Bottom 32 bits of counter
ints[1] = (shr32(x) & 0xFFFF) | DOUBLE_TOKEN_MARKER_BITS; // Top 16 bits + marker
return (new Float64Array(buf))[0];
```

### 4. **Key Properties**
- **Unique**: Each call generates a new unique number, even for the same token
- **Reversible**: The number can be decoded back to extract the original token
- **Detectable**: Special pattern allows detection of encoded tokens vs regular numbers
- **JSON-safe**: The encoded numbers survive JSON serialization/deserialization
- **Range**: Can encode up to 2^48-1 different token indices

### 5. **Example**
```typescript
const myToken = Token.asNumber(someResolvable);
// Returns something like: -1.2345678901234e+289
// This special number encodes the token's unique identifier
```

The generated numbers look like very large negative scientific notation numbers (e.g., `-1.1234567890123e+289`) but are actually specially crafted IEEE 754 doubles that encode the token's unique identifier in their bit representation.

---

Given this info, I think a conflict is very unlikely.